### PR TITLE
Fix: blueocean steps logs with logs mirroring

### PR DESCRIPTION
**Pull request by AndriiChuzhynov**
Using a decorator with FileLog storage prevents it from creating an index. Also, there is no way to write steps logs with decorator.

This change adds merged stream that takes two steams and applies all operations sequentially.

In this case, if the mirroring option is enabled. Both streams will be performed correctly

May be a fix for #563

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
Thanks @AndriiChuzhynov, I'll look at this ASAP, overall looks great

**Comment by cyrille-leclerc** [comment]
Sorry for the delay, this branch works well for me and the code makes sense to me.

**Comment by AndriiChuzhynov** [comment]
Cool! I'll finish the work and add the tests then. 
And convert it to the PR

**Comment by cyrille-leclerc** [comment]
@AndriiChuzhynov do you think we could move forward on this PR?
We have challenges with `OtelLocaLogMirroringTest` failing ([here](https://github.com/jenkinsci/opentelemetry-plugin/pull/603/checks?check_run_id=11959272841)) and I struggle to fix the problem. I would be happy that we work on the new code base :-)

**Comment by AndriiChuzhynov** [comment]
Hi, sure, I'll proceed with this PR on this weekends

**Comment by cyrille-leclerc** [comment]
Thank you very much Andrii!

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 1]
```diff
@@ -48,22 +50,35 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
...
+            File logFile = new File(buildFolderPath, "log");
+            return new MergedBuildListener(otelLogSenderBuildListenerOnController, FileLogStorage.forFile(logFile).overallListener());
+        }
+        return otelLogSenderBuildListenerOnController;
     }
 
     @Nonnull
     @Override
-    public TaskListener nodeListener(@Nonnull FlowNode flowNode) {
+    public TaskListener nodeListener(@Nonnull FlowNode flowNode) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:70]: I am concerned about these new Exceptions. They are throwing up without any explanation of what happen

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 2]
```diff
@@ -48,22 +50,35 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
 
     @Nonnull
     @Override
-    public BuildListener overallListener() {
+    public BuildListener overallListener() throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:53]: I am concerned about these new Exceptions. They are throwing up without any explanation of what happened

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 3]
```diff
@@ -48,22 +50,35 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
...
         Map<String, String> otelConfigurationProperties = otelConfiguration.toOpenTelemetryProperties();
         Map<String, String> otelResourceAttributes = new HashMap<>();
         otelConfiguration.toOpenTelemetryResource().getAttributes().asMap().forEach((k, v) -> otelResourceAttributes.put(k.getKey(), v.toString()));
-        return new OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController(buildInfo, otelConfigurationProperties, otelResourceAttributes);
+
+        OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController otelLogSenderBuildListenerOnController = new OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController(buildInfo, otelConfigurationProperties, otelResourceAttributes);
+
+        if (OpenTelemetrySdkProvider.get().isOtelLogsMirrorToDisk()) {
+            File logFile = new File(buildFolderPath, "log");
+            return new MergedBuildListener(otelLogSenderBuildListenerOnController, FileLogStorage.forFile(logFile).overallListener());
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:63]:
```suggestion
            try {
              File logFile = new File(buildFolderPath, "log");
              return new MergedBuildListener(otelLogSenderBuildListenerOnController, FileLogStorage.forFile(logFile).overallListener());
              catch (IOException|InterruptedException e) {
                throw new IOException("Was not possible to create the mirror logs.", e)
              }
```

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 4]
```diff
@@ -48,22 +50,35 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
...
         OpenTelemetryConfiguration otelConfiguration = JenkinsOpenTelemetryPluginConfiguration.get().toOpenTelemetryConfiguration();
         Map<String, String> otelConfigurationProperties = otelConfiguration.toOpenTelemetryProperties();
         Map<String, String> otelResourceAttributes = new HashMap<>();
         otelConfiguration.toOpenTelemetryResource().getAttributes().asMap().forEach((k, v) -> otelResourceAttributes.put(k.getKey(), v.toString()));
-        return new OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController(buildInfo, flowNode.getId(),  otelConfigurationProperties, otelResourceAttributes);
+        OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController otelLogSenderBuildListenerOnController = new OtelLogSenderBuildListener.OtelLogSenderBuildListenerOnController(buildInfo, flowNode.getId(), otelConfigurationProperties, otelResourceAttributes);
+
+        if (OpenTelemetrySdkProvider.get().isOtelLogsMirrorToDisk()) {
+            File logFile = new File(buildFolderPath, "log");
+            return new MergedTaskListener(otelLogSenderBuildListenerOnController, FileLogStorage.forFile(logFile).nodeListener(flowNode));
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:79]:
```suggestion
            try {
              File logFile = new File(buildFolderPath, "log");
              return new MergedTaskListener(otelLogSenderBuildListenerOnController, FileLogStorage.forFile(logFile).nodeListener(flowNode));
             } catch (IOException|InterruptedException e) {
                throw new IOException("Was not possible to create the mirror logs.", e)
              }
```

**Comment by kuisathaverat** [comment]
BTW BO is no longer maintained

**Comment by AndriiChuzhynov** [comment]
Changes applied, thanks @kuisathaverat

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 5]
```diff
@@ -48,22 +50,43 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
 
     @Nonnull
     @Override
-    public BuildListener overallListener() {
+    public BuildListener overallListener() throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:53]:
```suggestion
    public BuildListener overallListener() throws IOException {
```

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java [hunk 6]
```diff
@@ -48,22 +50,43 @@ public OtelLogStorage(@Nonnull BuildInfo buildInfo, @Nonnull Tracer tracer, @Nul
...
+                throw new IOException("Was not possible to create the mirror logs.", e);
+              }
+        }
+        return otelLogSenderBuildListenerOnController;
     }
 
     @Nonnull
     @Override
-    public TaskListener nodeListener(@Nonnull FlowNode flowNode) {
+    public TaskListener nodeListener(@Nonnull FlowNode flowNode) throws IOException, InterruptedException {
```
Comment (kuisathaverat) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:74]:
```suggestion
    public TaskListener nodeListener(@Nonnull FlowNode flowNode) throws IOException {
```

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java (hunk 1 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:74]: @kuisathaverat I'm not sure to catch your point. My understanding is that the `IOException` would be thrown by `org.jenkinsci.plugins.workflow.log.FileLogStorage#nodeListener` which sounds reasonable to me. We would have the same exceptions as when using standard log file backed logs.
On the `InterruptedException`, you are right it's not needed.

Did I miss something?

**Review by cyrille-leclerc — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java (hunk 1 above)
Comment (kuisathaverat) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java:74]: my comment was old, the exception is re-throw with a comment now that gives the user some info about what could happen.

**Review by cyrille-leclerc — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java (hunk 2 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogStorage.java]: Resolving as @kuisathaverat said in a comment I can no longer find that this comment is outdated.

**Review by kuisathaverat — APPROVED** [review]

**Pull request merged**